experimental/Schemes/Resolution_structure.jl

```diff
@@ -1,61 +1,201 @@
+
+  function MixedBlowUpSequence(maps::Vector{<:AbsCoveredSchemeMorphism})
+    n = length(maps)
+    for i in 1:n
+      @assert all(x->((x isa BlowupMorphism) || (x isa NormalizationMorphism)), maps) "only blow-ups and normalizations allowed"
+    end
+    for i in 1:n-1
+      @assert domain(maps[i]) === codomain(maps[i+1]) "not a sequence of morphisms"
+    end
+    resi = new{typeof(domain(maps[end])),typeof(codomain(first(maps)))}(maps)
+    resi.normalization_steps = [i for i in 1:n if maps[i] isa NormalizationMorphism]
+    return resi
+  end
+
+end
 
 
 ##################################################################################################
 # getters
 ##################################################################################################
+underlying_morphism(phi::NormalizationMorphism) = phi.underlying_morphism
+inclusion_morphisms(phi::NormalizationMorphism) = phi.inclusions
+normalization_steps(phi::NormalizationMorphism) = phi.normalization_steps
 morphisms(phi::AbsDesingMor) = copy(phi.maps)
-morphism(phi::AbsDesingMor,i::Int) = copy(phi.maps[i])
+morphism(phi::AbsDesingMor,i::Int) = phi.maps[i]
 last_map(phi::AbsDesingMor) = phi.maps[end]
-exceptional_divisor_list(phi::BlowUpSequence) = phi.ex_div  ## derzeit Liste von Eff. Cartier Div.
+
+exceptional_divisor_list(phi::BlowUpSequence) = phi.ex_div
+
+## entries of ex_div corresponding to normalization steps are only exceptional divisors at the very end
+## so only return them at the very end or on specific demand 
+function exceptional_divisor_list(phi::MixedBlowUpSequence, seq_unclean::Bool=false)
+  phi.resolves_sing || seq_unclean || error("excpetional divisor list not available for intermediate steps.")
+  return phi.ex_div
+end
+
+normalization_steps(phi::MixedBlowUpSequence) = phi.normalization_steps
+ 
 embeddings(phi::BlowUpSequence) = phi.embeddings
 
-## do not use!!! (for forwarding and certain emergenies)
 function underlying_morphism(phi::AbsDesingMor)
   if !isdefined(phi, :underlying_morphism)
     phi.underlying_morphism = CompositeCoveredSchemeMorphism(reverse(morphisms(phi)))
   end
   return phi.underlying_morphism
 end
 
+function exceptional_divisor(f::BlowUpSequence)
+  f.is_embedded && return _exceptional_divisor_in_ambient(f)
+  return _exceptional_divisor_non_embedded(f)
+end
+
+function exceptional_locus(phi::BlowUpSequence)
+  if !isdefined(phi, :exceptional_locus)
+    phi.exceptional_locus = weil_divisor(exceptional_divisor(phi))
+  end
+  return phi.exceptional_locus
+end
+
+@attr CartierDivisor function exceptional_divisor_with_multiplicities(f::BlowUpSequence)
+  is_embedded || error("only available for embedded desingularization")
+  phi.transform_type != :strict || error("only available for weak and controlled transforms")
+
+  ex_div_list = exceptional_divisor_list(f)
+  C = f.ex_mult[1] * cartier_divisor(ex_div_list[1])
+  for i in 2:length(ex_div_list)
+    C = C + f.ex_mult[i] * cartier_divisor(ex_div_list[i])
+  end
```

Comment:
```suggestion
  C = sum(f.ex_mult[i] * cartier_divisor(ex_div_list[i]) for i in 1:length(ex_div_list))
```

Reply:
Yes, after removing the debug output, I can now contract this to a single line.